lib/ecto/adapters/postgres/connection.ex

```diff
@@ -1556,15 +1556,34 @@ if Code.ensure_loaded?(Postgrex) do
     end
 
     defp column_change(table, {:modify, name, type, opts}) do
-      [
-        drop_reference_expr(opts[:from], table, name),
-        "ALTER COLUMN ",
-        quote_name(name),
-        " TYPE ",
-        column_type(type, opts),
-        modify_null(name, opts),
-        modify_default(name, type, opts)
-      ]
+      column_type = column_type(type, opts)
+
+      extract_type = fn
+        nil -> nil
+        {type, _} -> type
+        %Reference{} -> nil
+        type -> type
+      end
+
+      from_type = extract_type.(opts[:from])
+
+      if column_type == column_type(from_type, opts) do
+        [
+          drop_reference_expr(opts[:from], table, name),
+          modify_null(name, Keyword.put(opts, :prefix_with_comma, false)),
+          modify_default(name, type, opts)
+        ]
+      else
+        [
+          drop_reference_expr(opts[:from], table, name),
+          "ALTER COLUMN ",
+          quote_name(name),
+          " TYPE ",
+          column_type,
```

Comment:
Similar behavior is observed when doing something like:

```elixir
 modify :account_id, :binary_id, null: false, from: {:binary_id, null: true}
 ```

The SQL syntax produced is this one:

```sql
ALTER TABLE "my_table"
  ALTER COLUMN "account_id" TYPE uuid,
  ALTER COLUMN "account_id" SET NOT NULL;
```

And that's producing way more `AccessExclusiveLock` than the alternative: `ALTER TABLE "my_table" ALTER COLUMN "account_id" SET NOT NULL;`